File: src/main/java/cc/mirukuneko/realtrainmodrenewed/vehicle/VehicleDefinition.java
package cc.mirukuneko.realtrainmodrenewed.vehicle;

import net.minecraft.world.phys.Vec3;
import java.util.List;
import java.util.Map;
import java.util.ArrayList;

public class VehicleDefinition {
    public static final int SEAT_TYPE_DISABLED = 0;
    public static final int SEAT_TYPE_DRIVER_CAB = -1;

    public record BogieDefinition(String modelFile, Map<String, String> textureOverrides, Vec3 position, String scriptPath) {
        public BogieDefinition(String modelFile, Map<String, String> textureOverrides, Vec3 position) {
            this(modelFile, textureOverrides, position, "");
        }
    }

    public record SeatMarker(Vec3 position, int type, boolean driverCab) {
        public boolean isRideable() {
            return this.driverCab || this.type != SEAT_TYPE_DISABLED;
        }
    }

    public record DoorAnimationDefinition(List<String> objects, Vec3 closedPosition, Vec3 openTranslation) {
    }

    public record RollsignDefinition(float[] uv, float[][][] pos, boolean doAnimation, boolean disableLighting) {
    }

    public record LightDefinition(byte type, int color, Vec3 position, float radius, boolean reverse) {
    }

    private final String id;
    private final String displayName;
    private final String packName;
    private final String modelFile;
    private final String buttonTexture;
    private final Map<String, String> textureOverrides;
    private final Vec3 modelOffset;
    private final float modelScale;
    private final List<BogieDefinition> bogies;
    private final List<SeatMarker> seatMarkers;
    private final List<Vec3> seatPositions;
    private final List<Vec3> playerPositions;
    private final Vec3 seatOffset;
    private final String scriptPath;
    private final String soundScriptPath;
    // SuperRailBuilder3 などのサーバ側スクリプト（オプション。後付け設定）
    private String serverScriptPath = "";
    private final String vehicleType;
    private final String doorType;
    private final float trainDistance;
    private final int driverSeatIndex;
    private final int frontDriverSeatIndex;
    private final int rearDriverSeatIndex;
    private final List<DoorAnimationDefinition> leftDoors;
    private final List<DoorAnimationDefinition> rightDoors;
    private final List<Float> notchMaxSpeeds;
    private final float acceleration;
    private final boolean smoothing;
    private final List<String> rollsignNames;
    private final List<String> customButtonNames;
    private final List<List<String>> customButtonOptions;
    private final String rollsignTexture;
    private final List<RollsignDefinition> rollsigns;
    private final List<LightDefinition> headLights;
    private final List<LightDefinition> tailLights;
    private final List<LightDefinition> interiorLights;
    private final String hornSound;
    private final List<String> announcementSounds;
    private String soundStop = "";
    private String soundStartAcceleration = "";
    private String soundAcceleration = "";
    private String soundDeceleration = "";
    private String soundDecelerationStop = "";
    private final boolean doCulling;
    private final boolean renderLight;
    private final boolean notDisplayCab;
    private final boolean singleTrain;

    public VehicleDefinition(
        String id,
        String displayName,
        String packName,
        String modelFile,
        String buttonTexture,
        Map<String, String> textureOverrides,
        Vec3 modelOffset,
        float modelScale,
        List<BogieDefinition> bogies,
        List<Vec3> seatPositions,
        List<Vec3> playerPositions,
        Vec3 seatOffset,
        String scriptPath,
        String soundScriptPath,
        String vehicleType,
        String doorType,
        float trainDistance,
        int driverSeatIndex,
        int frontDriverSeatIndex,
        int rearDriverSeatIndex,
        List<DoorAnimationDefinition> leftDoors,
        List<DoorAnimationDefinition> rightDoors,
        List<Float> notchMaxSpeeds,
        float acceleration,
        boolean smoothing,
        List<String> rollsignNames,
        List<String> customButtonNames,
        List<List<String>> customButtonOptions,
        String rollsignTexture,
        List<RollsignDefinition> rollsigns,
        List<LightDefinition> headLights,
        List<LightDefinition> tailLights,
        List<LightDefinition> interiorLights,
        String hornSound,
        List<String> announcementSounds,
        boolean doCulling,
        boolean renderLight,
        boolean notDisplayCab,
        boolean singleTrain
    ) {
        this(
            id,
            displayName,
            packName,
            modelFile,
            buttonTexture,
            textureOverrides,
            modelOffset,
            modelScale,
            bogies,
            buildSeatMarkers(seatPositions, playerPositions),
            seatPositions,
            playerPositions,
            seatOffset,
            scriptPath,
            soundScriptPath,
            vehicleType,
            doorType,
            trainDistance,
            driverSeatIndex,
            frontDriverSeatIndex,
            rearDriverSeatIndex,
            leftDoors,
            rightDoors,
            notchMaxSpeeds,
            acceleration,
            smoothing,
            rollsignNames,
            customButtonNames,
            customButtonOptions,
            rollsignTexture,
            rollsigns,
            headLights,
            tailLights,
            interiorLights,
            hornSound,
            announcementSounds,
            doCulling,
            renderLight,
            notDisplayCab,
            singleTrain
        );
    }

    public VehicleDefinition(
        String id,
        String displayName,
        String packName,
        String modelFile,
        String buttonTexture,
        Map<String, String> textureOverrides,
        Vec3 modelOffset,
        float modelScale,
        List<BogieDefinition> bogies,
        List<SeatMarker> seatMarkers,
        List<Vec3> seatPositions,
        List<Vec3> playerPositions,
        Vec3 seatOffset,
        String scriptPath,
        String soundScriptPath,
        String vehicleType,
        String doorType,
        float trainDistance,
        int driverSeatIndex,
        int frontDriverSeatIndex,
        int rearDriverSeatIndex,
        List<DoorAnimationDefinition> leftDoors,
        List<DoorAnimationDefinition> rightDoors,
        List<Float> notchMaxSpeeds,
        float acceleration,
        boolean smoothing,
        List<String> rollsignNames,
        List<String> customButtonNames,
        List<List<String>> customButtonOptions,
        String rollsignTexture,
        List<RollsignDefinition> rollsigns,
        List<LightDefinition> headLights,
        List<LightDefinition> tailLights,
        List<LightDefinition> interiorLights,
        String hornSound,
        List<String> announcementSounds,
        boolean doCulling,
        boolean renderLight,
        boolean notDisplayCab,
        boolean singleTrain
    ) {
        this.id = id;
        this.displayName = displayName;
        this.packName = packName;
        this.modelFile = modelFile;
        this.buttonTexture = buttonTexture == null ? "" : buttonTexture;
        this.textureOverrides = textureOverrides == null ? Map.of() : Map.copyOf(textureOverrides);
        this.modelOffset = modelOffset == null ? Vec3.ZERO : modelOffset;
        this.modelScale = modelScale <= 0 ? 1.0F : modelScale;
        this.bogies = bogies == null ? List.of() : List.copyOf(bogies);
        this.seatMarkers = seatMarkers == null ? List.of() : List.copyOf(seatMarkers);
        this.seatPositions = seatPositions == null ? List.of() : List.copyOf(seatPositions);
        this.playerPositions = playerPositions == null ? List.of() : List.copyOf(playerPositions);
        this.seatOffset = seatOffset;
        this.scriptPath = scriptPath == null ? "" : scriptPath;
        this.soundScriptPath = soundScriptPath == null ? "" : soundScriptPath;
        this.vehicleType = vehicleType == null ? "Train" : vehicleType;
        this.doorType = doorType == null ? "manual" : doorType;
        this.trainDistance = trainDistance > 0 ? trainDistance : 4.5F;
        this.driverSeatIndex = driverSeatIndex;
        this.frontDriverSeatIndex = frontDriverSeatIndex;
        this.rearDriverSeatIndex = rearDriverSeatIndex;
        this.leftDoors = leftDoors == null ? List.of() : List.copyOf(leftDoors);
        this.rightDoors = rightDoors == null ? List.of() : List.copyOf(rightDoors);
        this.notchMaxSpeeds = notchMaxSpeeds == null ? List.of() : List.copyOf(notchMaxSpeeds);
        this.acceleration = acceleration > 0 ? acceleration : 0.00243F;
        this.smoothing = smoothing;
        this.rollsignNames = rollsignNames == null ? List.of() : List.copyOf(rollsignNames);
        this.customButtonNames = customButtonNames == null ? List.of() : List.copyOf(customButtonNames);
        this.customButtonOptions = customButtonOptions == null ? List.of() : toImmutableNestedList(customButtonOptions);
        this.rollsignTexture = rollsignTexture == null ? "" : rollsignTexture;
        this.rollsigns = rollsigns == null ? List.of() : List.copyOf(rollsigns);
        this.headLights = headLights == null ? List.of() : List.copyOf(headLights);
        this.tailLights = tailLights == null ? List.of() : List.copyOf(tailLights);
        this.interiorLights = interiorLights == null ? List.of() : List.copyOf(interiorLights);
        this.hornSound = hornSound == null ? "" : hornSound;
        this.announcementSounds = announcementSounds == null ? List.of() : List.copyOf(announcementSounds);
        this.doCulling = doCulling;
        this.renderLight = renderLight;
        this.notDisplayCab = notDisplayCab;
        this.singleTrain = singleTrain;
    }

    public String getId() { return id; }
    public String getDisplayName() { return displayName; }
    public String getPackName() { return packName; }
    public String getModelFile() { return modelFile; }
    public String getButtonTexture() { return buttonTexture; }
    public Map<String, String> getTextureOverrides() { return textureOverrides; }
    public Vec3 getModelOffset() { return modelOffset; }
    public float getModelScale() { return modelScale; }
    public List<BogieDefinition> getBogies() { return bogies; }
    public List<Vec3> getBogiePositions() { 
        List<Vec3> positions = new ArrayList<>();
        for (BogieDefinition bogie : bogies) {
            positions.add(bogie.position);
        }
        return positions;
    }
    public List<Vec3> getSeatPositions() { return seatPositions; }
    public List<Vec3> getPlayerPositions() { return playerPositions; }
    public List<SeatMarker> getSeatMarkers() { return seatMarkers; }

    public List<SeatMarker> getRideableSeatMarkers() {
        if (seatMarkers.isEmpty()) {
            return List.of();
        }
        List<SeatMarker> rideable = new ArrayList<>(seatMarkers.size());
        for (SeatMarker marker : seatMarkers) {
            if (marker.isRideable()) {
                rideable.add(marker);
            }
        }
        return List.copyOf(rideable);
    }

    public boolean isDriverSeatIndex(int index) {
        List<SeatMarker> rideable = getRideableSeatMarkers();
        return index >= 0 && index < rideable.size() && rideable.get(index).driverCab();
    }

    public List<Vec3> getAllSeatPositions() {
        if (seatMarkers.isEmpty()) {
            if (playerPositions.isEmpty()) {
                return seatPositions;
            }
            if (seatPositions.isEmpty()) {
                return playerPositions;
            }
            List<Vec3> seats = new ArrayList<>(playerPositions.size() + seatPositions.size());
            seats.addAll(playerPositions);
            seats.addAll(seatPositions);
            return List.copyOf(seats);
        }
        List<Vec3> seats = new ArrayList<>(seatMarkers.size());
        for (SeatMarker marker : seatMarkers) {
            seats.add(marker.position());
        }
        return List.copyOf(seats);
    }

    public List<Vec3> getRideableSeatPositions() {
        List<SeatMarker> rideable = getRideableSeatMarkers();
        if (rideable.isEmpty()) {
            return List.of();
        }
        List<Vec3> positions = new ArrayList<>(rideable.size());
        for (SeatMarker marker : rideable) {
            positions.add(marker.position());
        }
        return List.copyOf(positions);
    }

    public boolean hasSeatOffset() {
        return seatOffset != null;
    }

    public Vec3 getSeatOffset() {
        return seatOffset == null ? Vec3.ZERO : seatOffset;
    }

    public boolean hasScript() {
        return scriptPath != null && !scriptPath.isBlank();
    }

    public String getScriptPath() {
        return scriptPath;
    }

    public boolean hasSoundScript() {
        return soundScriptPath != null && !soundScriptPath.isBlank();
    }

    public String getSoundScriptPath() {
        return soundScriptPath;
    }

    public String getServerScriptPath() {
        return serverScriptPath;
    }

    public boolean hasServerScript() {
        return serverScriptPath != null && !serverScriptPath.isBlank();
    }

    /** Loader が JSON 読込後に設定する用。コンストラクタ大量改修を避けるため後付け。 */
    public void setServerScriptPath(String path) {
        this.serverScriptPath = path == null ? "" : path;
    }

    public String getVehicleType() {
        return vehicleType;
    }

    public boolean isCarType() {
        return "car".equalsIgnoreCase(vehicleType);
    }

    public String getDoorType() {
        return doorType;
    }

    public boolean hasAutomaticDoor() {
        return "automatic".equalsIgnoreCase(doorType) || "auto".equalsIgnoreCase(doorType);
    }

    public float getTrainDistance() {
        return trainDistance;
    }

    public int getDriverSeatIndex() {
        return driverSeatIndex;
    }

    public int getFrontDriverSeatIndex() {
        return frontDriverSeatIndex;
    }

    public int getRearDriverSeatIndex() {
        return rearDriverSeatIndex;
    }

    public List<DoorAnimationDefinition> getLeftDoors() {
        return leftDoors;
    }

    public List<DoorAnimationDefinition> getRightDoors() {
        return rightDoors;
    }

    public boolean hasDoorAnimations() {
        return !leftDoors.isEmpty() || !rightDoors.isEmpty();
    }

    public List<Float> getNotchMaxSpeeds() {
        return notchMaxSpeeds;
    }

    public float getAcceleration() {
        return acceleration;
    }

    public boolean isSmoothing() {
        return smoothing;
    }

    public List<String> getRollsignNames() {
        return rollsignNames;
    }

    public List<String> getCustomButtonNames() {
        return customButtonNames;
    }

    public List<List<String>> getCustomButtonOptions() {
        return customButtonOptions;
    }

    public String getRollsignTexture() {
        return rollsignTexture;
    }

    public List<RollsignDefinition> getRollsigns() {
        return rollsigns;
    }

    public List<LightDefinition> getHeadLights() {
        return headLights;
    }

    public List<LightDefinition> getTailLights() {
        return tailLights;
    }

    public List<LightDefinition> getInteriorLights() {
        return interiorLights;
    }

    public String getHornSound() {
        return hornSound;
    }

    public List<String> getAnnouncementSounds() {
        return announcementSounds;
    }

    public String getSoundStop() {
        return soundStop;
    }

    public String getSoundStartAcceleration() {
        return soundStartAcceleration;
    }

    public String getSoundAcceleration() {
        return soundAcceleration;
    }

    public String getSoundDeceleration() {
        return soundDeceleration;
    }

    public String getSoundDecelerationStop() {
        return soundDecelerationStop;
    }

    public boolean hasJsonRunningSounds() {
        return !soundStop.isBlank()
            || !soundStartAcceleration.isBlank()
            || !soundAcceleration.isBlank()
            || !soundDeceleration.isBlank()
            || !soundDecelerationStop.isBlank();
    }

    public void setJsonRunningSounds(String soundStop, String soundStartAcceleration,
                                     String soundAcceleration, String soundDeceleration,
                                     String soundDecelerationStop) {
        this.soundStop = soundStop == null ? "" : soundStop;
        this.soundStartAcceleration = soundStartAcceleration == null ? "" : soundStartAcceleration;
        this.soundAcceleration = soundAcceleration == null ? "" : soundAcceleration;
        this.soundDeceleration = soundDeceleration == null ? "" : soundDeceleration;
        this.soundDecelerationStop = soundDecelerationStop == null ? "" : soundDecelerationStop;
    }

    public boolean isDoCulling() {
        return doCulling;
    }

    public boolean isRenderLight() {
        return renderLight;
    }

    public boolean isNotDisplayCab() {
        return notDisplayCab;
    }

    public boolean isSingleTrain() {
        return singleTrain;
    }

    private static List<List<String>> toImmutableNestedList(List<List<String>> options) {
        if (options.isEmpty()) {
            return List.of();
        }
        List<List<String>> copy = new ArrayList<>(options.size());
        for (List<String> option : options) {
            copy.add(option == null ? List.of() : List.copyOf(option));
        }
        return List.copyOf(copy);
    }

    private static List<SeatMarker> buildSeatMarkers(List<Vec3> seatPositions, List<Vec3> playerPositions) {
        List<SeatMarker> markers = new ArrayList<>();
        if (playerPositions != null) {
            for (Vec3 playerPos : playerPositions) {
                markers.add(new SeatMarker(playerPos, SEAT_TYPE_DRIVER_CAB, true));
            }
        }
        if (seatPositions != null) {
            for (Vec3 seatPos : seatPositions) {
                markers.add(new SeatMarker(seatPos, 1, false));
            }
        }
        return List.copyOf(markers);
    }
}

File: src/main/kotlin/cc/mirukuneko/realtrainmodrenewed/vehicle/VehicleDefinition.kt
// SPDX-License-Identifier: LGPL-3.0-or-later
// Copyright © 2026 mirukuneko and RealTrainModRenewed contributors
package cc.mirukuneko.realtrainmodrenewed.vehicle

import net.minecraft.world.phys.Vec3

class VehicleDefinition(
    @JvmField val id: String,
    @JvmField val displayName: String,
    @JvmField val packName: String,
    @JvmField val modelFile: String,
    buttonTexture: String?,
    textureOverrides: Map<String, String>?,
    modelOffset: Vec3?,
    modelScale: Float,
    @JvmField val bogies: List<BogieDefinition>,
    @JvmField val seatMarkers: List<SeatMarker>,
    @JvmField val seatPositions: List<Vec3>,
    @JvmField val playerPositions: List<Vec3>,
    seatOffset: Vec3?,
    @JvmField val scriptPath: String,
    @JvmField val soundScriptPath: String,
    @JvmField var vehicleType: String,
    @JvmField val doorType: String,
    @JvmField val trainDistance: Float,
    @JvmField val driverSeatIndex: Int,
    @JvmField val frontDriverSeatIndex: Int,
    @JvmField val rearDriverSeatIndex: Int,
    @JvmField val leftDoors: List<DoorAnimationDefinition>,
    @JvmField val rightDoors: List<DoorAnimationDefinition>,
    @JvmField val notchMaxSpeeds: List<Float>,
    @JvmField val brakeDecelerations: List<Float>,
    @JvmField val acceleration: Float,
    @JvmField val smoothing: Boolean,
    @JvmField val rollsignNames: List<String>,
    @JvmField val customButtonNames: List<String>,
    @JvmField val customButtonOptions: List<List<String>>,
    @JvmField val rollsignTexture: String,
    @JvmField val rollsigns: List<RollsignDefinition>,
    @JvmField val headLights: List<LightDefinition>,
    @JvmField val tailLights: List<LightDefinition>,
    @JvmField val interiorLights: List<LightDefinition>,
    @JvmField val hornSound: String,
    @JvmField val announcementSounds: List<String>,
    @JvmField val doCulling: Boolean,
    @JvmField val renderLight: Boolean,
    @JvmField val notDisplayCab: Boolean,
    @JvmField val singleTrain: Boolean
) {
    data class BogieDefinition(
        @JvmField val modelFile: String,
        @JvmField val textureOverrides: Map<String, String>,
        @JvmField val position: Vec3,
        @JvmField val scriptPath: String
    ) {
        constructor(modelFile: String, textureOverrides: Map<String, String>, position: Vec3) :
            this(modelFile, textureOverrides, position, "")

        fun getModelFile(): String = modelFile
        fun getTextureOverrides(): Map<String, String> = textureOverrides
        fun getPosition(): Vec3 = position
        fun getScriptPath(): String = scriptPath

        fun modelFile(): String = modelFile
        fun textureOverrides(): Map<String, String> = textureOverrides
        fun position(): Vec3 = position
        fun scriptPath(): String = scriptPath
    }

    data class SeatMarker(
        @JvmField val position: Vec3,
        @JvmField val type: Int,
        @JvmField val driverCab: Boolean
    ) {
        fun getPosition(): Vec3 = position
        fun getType(): Int = type
        fun isDriverCab(): Boolean = driverCab
        fun position(): Vec3 = position
        fun type(): Int = type
        fun driverCab(): Boolean = driverCab

        fun isRideable(): Boolean = driverCab || type != SEAT_TYPE_DISABLED
    }

    data class DoorAnimationDefinition(
        @JvmField val objects: List<String>,
        @JvmField val closedPosition: Vec3,
        @JvmField val openTranslation: Vec3
    ) {
        fun getObjects(): List<String> = objects
        fun getClosedPosition(): Vec3 = closedPosition
        fun getOpenTranslation(): Vec3 = openTranslation
        fun objects(): List<String> = objects
        fun closedPosition(): Vec3 = closedPosition
        fun openTranslation(): Vec3 = openTranslation
    }

    data class RollsignDefinition(
        @JvmField val uv: FloatArray,
        @JvmField val pos: Array<Array<FloatArray>>,
        @JvmField val doAnimation: Boolean,
        @JvmField val disableLighting: Boolean
    ) {
        fun getUv(): FloatArray = uv
        fun getPos(): Array<Array<FloatArray>> = pos
        fun isDoAnimation(): Boolean = doAnimation
        fun isDisableLighting(): Boolean = disableLighting
        fun uv(): FloatArray = uv
        fun pos(): Array<Array<FloatArray>> = pos
        fun doAnimation(): Boolean = doAnimation
        fun disableLighting(): Boolean = disableLighting

        override fun equals(other: Any?): Boolean {
            if (this === other) return true
            if (other !is RollsignDefinition) return false
            return uv.contentEquals(other.uv) && pos.contentDeepEquals(other.pos) &&
                doAnimation == other.doAnimation && disableLighting == other.disableLighting
        }
        override fun hashCode(): Int {
            var result = uv.contentHashCode()
            result = 31 * result + pos.contentDeepHashCode()
            result = 31 * result + doAnimation.hashCode()
            result = 31 * result + disableLighting.hashCode()
            return result
        }
    }

    data class LightDefinition(
        @JvmField val type: Byte,
        @JvmField val color: Int,
        @JvmField val position: Vec3,
        @JvmField val radius: Float,
        @JvmField val reverse: Boolean
    ) {
        fun getType(): Byte = type
        fun getColor(): Int = color
        fun getPosition(): Vec3 = position
        fun getRadius(): Float = radius
        fun isReverse(): Boolean = reverse
        fun type(): Byte = type
        fun color(): Int = color
        fun position(): Vec3 = position
        fun radius(): Float = radius
        fun reverse(): Boolean = reverse
    }

    companion object {
        const val SEAT_TYPE_DISABLED: Int = 0
        const val SEAT_TYPE_DRIVER_CAB: Int = -1

        private fun buildSeatMarkers(seatPositions: List<Vec3>?, playerPositions: List<Vec3>?): List<SeatMarker> {
            val markers = mutableListOf<SeatMarker>()
            playerPositions?.forEach { markers.add(SeatMarker(it, SEAT_TYPE_DRIVER_CAB, true)) }
            seatPositions?.forEach { markers.add(SeatMarker(it, 1, false)) }
            return markers.toList()
        }

        private fun toImmutableNestedList(options: List<List<String>>?): List<List<String>> {
            if (options.isNullOrEmpty()) return emptyList()
            return options.map { it.toList() }.toList()
        }
    }

    @JvmField val buttonTexture: String = buttonTexture ?: ""
    @JvmField val textureOverrides: Map<String, String> = textureOverrides ?: emptyMap()
    @JvmField val modelOffset: Vec3 = modelOffset ?: Vec3.ZERO
    @JvmField val modelScale: Float = if (modelScale <= 0.0F) 1.0F else modelScale
    @JvmField val seatOffset: Vec3 = seatOffset ?: Vec3.ZERO

    @JvmField var serverScriptPath: String = ""
    @JvmField var doorOpenSound: String = ""
    @JvmField var doorCloseSound: String = ""
    @JvmField var soundStop: String = ""
    @JvmField var soundStartAcceleration: String = ""
    @JvmField var soundAcceleration: String = ""
    @JvmField var soundDeceleration: String = ""
    @JvmField var soundDecelerationStop: String = ""
    @JvmField var announcementNames: List<String> = emptyList()
    @JvmField var typeSignNames: List<String> = emptyList()
    @JvmField var typeSignTexture: String = ""
    @JvmField var typeSigns: List<RollsignDefinition> = emptyList()
    @JvmField var notchAccelerations: List<Float> = emptyList()
    @JvmField var useVariableAcceleration: Boolean = false
    @JvmField var useVariableDeceleration: Boolean = false

    fun getId(): String = id
    fun getDisplayName(): String = displayName
    fun getPackName(): String = packName
    fun getModelFile(): String = modelFile
    fun getButtonTexture(): String = buttonTexture
    fun getTextureOverrides(): Map<String, String> = textureOverrides
    fun getModelOffset(): Vec3 = modelOffset
    fun getModelScale(): Float = modelScale
    fun getBogies(): List<BogieDefinition> = bogies
    fun getSeatMarkers(): List<SeatMarker> = seatMarkers
    fun getSeatPositions(): List<Vec3> = seatPositions
    fun getPlayerPositions(): List<Vec3> = playerPositions
    fun getSeatOffset(): Vec3 = seatOffset
    fun getScriptPath(): String = scriptPath
    fun getSoundScriptPath(): String = soundScriptPath
    fun getVehicleType(): String = vehicleType
    fun getDoorType(): String = doorType
    fun getTrainDistance(): Float = trainDistance
    fun getDriverSeatIndex(): Int = driverSeatIndex
    fun getFrontDriverSeatIndex(): Int = frontDriverSeatIndex
    fun getRearDriverSeatIndex(): Int = rearDriverSeatIndex
    fun getLeftDoors(): List<DoorAnimationDefinition> = leftDoors
    fun getRightDoors(): List<DoorAnimationDefinition> = rightDoors
    fun getNotchMaxSpeeds(): List<Float> = notchMaxSpeeds
    fun getBrakeDecelerations(): List<Float> = brakeDecelerations
    fun getAcceleration(): Float = acceleration
    fun getNotchAccelerations(): List<Float> = notchAccelerations
    fun isUseVariableAcceleration(): Boolean = useVariableAcceleration
    fun isUseVariableDeceleration(): Boolean = useVariableDeceleration
    fun isSmoothing(): Boolean = smoothing
    fun getRollsignNames(): List<String> = rollsignNames
    fun getCustomButtonNames(): List<String> = customButtonNames
    fun getCustomButtonOptions(): List<List<String>> = customButtonOptions
    fun getRollsignTexture(): String = rollsignTexture
    fun getRollsigns(): List<RollsignDefinition> = rollsigns
    fun getHeadLights(): List<LightDefinition> = headLights
    fun getTailLights(): List<LightDefinition> = tailLights
    fun getInteriorLights(): List<LightDefinition> = interiorLights
    fun getHornSound(): String = hornSound
    fun getAnnouncementSounds(): List<String> = announcementSounds
    fun getAnnouncementNames(): List<String> = announcementNames
    fun getTypeSignNames(): List<String> = typeSignNames
    fun getTypeSignTexture(): String = typeSignTexture
    fun getTypeSigns(): List<RollsignDefinition> = typeSigns
    fun isDoCulling(): Boolean = doCulling
    fun isRenderLight(): Boolean = renderLight
    fun isNotDisplayCab(): Boolean = notDisplayCab
    fun isSingleTrain(): Boolean = singleTrain

    fun getBogiePositions(): List<Vec3> = bogies.map { it.position }

    fun getRideableSeatMarkers(): List<SeatMarker> = seatMarkers.filter { it.isRideable() }

    fun isDriverSeatIndex(index: Int): Boolean {
        val rideable = getRideableSeatMarkers()
        return index >= 0 && index < rideable.size && rideable[index].driverCab
    }

    fun getAllSeatPositions(): List<Vec3> {
        if (seatMarkers.isNotEmpty()) return seatMarkers.map { it.position }
        return when {
            playerPositions.isEmpty() -> seatPositions
            seatPositions.isEmpty() -> playerPositions
            else -> playerPositions + seatPositions
        }
    }

    fun getRideableSeatPositions(): List<Vec3> =
        getRideableSeatMarkers().map { it.position }

    fun hasSeatOffset(): Boolean = seatOffset != Vec3.ZERO
    fun hasScript(): Boolean = scriptPath.isNotBlank()
    fun hasSoundScript(): Boolean = soundScriptPath.isNotBlank()
    fun getServerScriptPath(): String = serverScriptPath
    fun hasServerScript(): Boolean = serverScriptPath.isNotBlank()

    fun setServerScriptPath(path: String?) { serverScriptPath = path ?: "" }

    fun setAnnouncementNames(names: List<String>?) {
        announcementNames = names?.toList() ?: emptyList()
    }

    fun setTypeSign(names: List<String>?, texture: String?, signs: List<RollsignDefinition>?) {
        typeSignNames = names?.toList() ?: emptyList()
        typeSignTexture = texture ?: ""
        typeSigns = signs?.toList() ?: emptyList()
    }

    fun isCarType(): Boolean = vehicleType.equals("car", ignoreCase = true)
    fun hasAutomaticDoor(): Boolean =
        doorType.equals("automatic", ignoreCase = true) || doorType.equals("auto", ignoreCase = true)
    fun hasDoorAnimations(): Boolean = leftDoors.isNotEmpty() || rightDoors.isNotEmpty()

    fun getDoorOpenSound(): String = doorOpenSound
    fun getDoorCloseSound(): String = doorCloseSound

    fun setDoorSounds(open: String?, close: String?) {
        doorOpenSound = open ?: ""
        doorCloseSound = close ?: ""
    }

    fun getSoundStop(): String = soundStop
    fun getSoundStartAcceleration(): String = soundStartAcceleration
    fun getSoundAcceleration(): String = soundAcceleration
    fun getSoundDeceleration(): String = soundDeceleration
    fun getSoundDecelerationStop(): String = soundDecelerationStop

    fun hasJsonRunningSounds(): Boolean =
        soundStop.isNotBlank() || soundStartAcceleration.isNotBlank() ||
        soundAcceleration.isNotBlank() || soundDeceleration.isNotBlank() ||
        soundDecelerationStop.isNotBlank()

    fun setJsonRunningSounds(stop: String?, startAccel: String?, accel: String?, decel: String?, decelStop: String?) {
        soundStop = stop ?: ""
        soundStartAcceleration = startAccel ?: ""
        soundAcceleration = accel ?: ""
        soundDeceleration = decel ?: ""
        soundDecelerationStop = decelStop ?: ""
    }

    // Simplified constructors matching Java overloads
    constructor(
        id: String, displayName: String, packName: String, modelFile: String,
        buttonTexture: String?, textureOverrides: Map<String, String>?, modelOffset: Vec3?,
        modelScale: Float, bogies: List<BogieDefinition>, seatPositions: List<Vec3>,
        playerPositions: List<Vec3>, seatOffset: Vec3?, scriptPath: String, soundScriptPath: String,
        vehicleType: String, doorType: String, trainDistance: Float, driverSeatIndex: Int,
        frontDriverSeatIndex: Int, rearDriverSeatIndex: Int, leftDoors: List<DoorAnimationDefinition>,
        rightDoors: List<DoorAnimationDefinition>, notchMaxSpeeds: List<Float>,
        acceleration: Float, smoothing: Boolean, rollsignNames: List<String>,
        customButtonNames: List<String>, customButtonOptions: List<List<String>>?,
        rollsignTexture: String, rollsigns: List<RollsignDefinition>, headLights: List<LightDefinition>,
        tailLights: List<LightDefinition>, interiorLights: List<LightDefinition>,
        hornSound: String, announcementSounds: List<String>, doCulling: Boolean,
        renderLight: Boolean, notDisplayCab: Boolean, singleTrain: Boolean
    ) : this(id, displayName, packName, modelFile, buttonTexture, textureOverrides, modelOffset, modelScale,
        bogies, buildSeatMarkers(seatPositions, playerPositions), seatPositions, playerPositions,
        seatOffset, scriptPath, soundScriptPath, vehicleType, doorType, trainDistance,
        driverSeatIndex, frontDriverSeatIndex, rearDriverSeatIndex, leftDoors, rightDoors,
        notchMaxSpeeds, emptyList(), acceleration, smoothing, rollsignNames,
        customButtonNames, toImmutableNestedList(customButtonOptions), rollsignTexture, rollsigns,
        headLights, tailLights, interiorLights, hornSound, announcementSounds,
        doCulling, renderLight, notDisplayCab, singleTrain)

    constructor(
        id: String, displayName: String, packName: String, modelFile: String,
        buttonTexture: String?, textureOverrides: Map<String, String>?, modelOffset: Vec3?,
        modelScale: Float, bogies: List<BogieDefinition>, seatPositions: List<Vec3>,
        playerPositions: List<Vec3>, seatOffset: Vec3?, scriptPath: String, soundScriptPath: String,
        vehicleType: String, doorType: String, trainDistance: Float, driverSeatIndex: Int,
        frontDriverSeatIndex: Int, rearDriverSeatIndex: Int, leftDoors: List<DoorAnimationDefinition>,
        rightDoors: List<DoorAnimationDefinition>, notchMaxSpeeds: List<Float>,
        brakeDecelerations: List<Float>, acceleration: Float, smoothing: Boolean,
        rollsignNames: List<String>, customButtonNames: List<String>,
        customButtonOptions: List<List<String>>?, rollsignTexture: String,
        rollsigns: List<RollsignDefinition>, headLights: List<LightDefinition>,
        tailLights: List<LightDefinition>, interiorLights: List<LightDefinition>,
        hornSound: String, announcementSounds: List<String>, doCulling: Boolean,
        renderLight: Boolean, notDisplayCab: Boolean, singleTrain: Boolean
    ) : this(id, displayName, packName, modelFile, buttonTexture, textureOverrides, modelOffset, modelScale,
        bogies, buildSeatMarkers(seatPositions, playerPositions), seatPositions, playerPositions,
        seatOffset, scriptPath, soundScriptPath, vehicleType, doorType, trainDistance,
        driverSeatIndex, frontDriverSeatIndex, rearDriverSeatIndex, leftDoors, rightDoors,
        notchMaxSpeeds, brakeDecelerations, acceleration, smoothing, rollsignNames,
        customButtonNames, toImmutableNestedList(customButtonOptions), rollsignTexture, rollsigns,
        headLights, tailLights, interiorLights, hornSound, announcementSounds,
        doCulling, renderLight, notDisplayCab, singleTrain)

    // Intentionally same signature as below - second overload for simpler usage
    @Suppress("UNUSED_PARAMETER")
    constructor(
        _unused: Unit = Unit,  // disambiguator - first of the simplified constructors without brakeDecelerations
        id: String, displayName: String, packName: String, modelFile: String,
        buttonTexture: String?, textureOverrides: Map<String, String>?, modelOffset: Vec3?,
        modelScale: Float, bogies: List<BogieDefinition>, seatMarkers: List<SeatMarker>,
        seatPositions: List<Vec3>, playerPositions: List<Vec3>, seatOffset: Vec3?,
        scriptPath: String, soundScriptPath: String, vehicleType: String, doorType: String,
        trainDistance: Float, driverSeatIndex: Int, frontDriverSeatIndex: Int,
        rearDriverSeatIndex: Int, leftDoors: List<DoorAnimationDefinition>,
        rightDoors: List<DoorAnimationDefinition>, notchMaxSpeeds: List<Float>,
        acceleration: Float, smoothing: Boolean, rollsignNames: List<String>,
        customButtonNames: List<String>, customButtonOptions: List<List<String>>?,
        rollsignTexture: String, rollsigns: List<RollsignDefinition>, headLights: List<LightDefinition>,
        tailLights: List<LightDefinition>, interiorLights: List<LightDefinition>,
        hornSound: String, announcementSounds: List<String>, doCulling: Boolean,
        renderLight: Boolean, notDisplayCab: Boolean, singleTrain: Boolean
    ) : this(id, displayName, packName, modelFile, buttonTexture, textureOverrides, modelOffset, modelScale,
        bogies, seatMarkers, seatPositions, playerPositions, seatOffset, scriptPath, soundScriptPath,
        vehicleType, doorType, trainDistance, driverSeatIndex, frontDriverSeatIndex, rearDriverSeatIndex,
        leftDoors, rightDoors, notchMaxSpeeds, emptyList(), acceleration, smoothing, rollsignNames,
        customButtonNames, toImmutableNestedList(customButtonOptions), rollsignTexture, rollsigns,
        headLights, tailLights, interiorLights, hornSound, announcementSounds,
        doCulling, renderLight, notDisplayCab, singleTrain)
}

